### ISSUE TITLE
System.URI fails to correctly parse URNs containing unicode characters

### ISSUE THREAD

**Issue by richardprobert** [dotnet/corefx#23783]
System.URI fails to correctly parse URNs containing unicode characters. For reproduction steps, see this Gist: https://gist.github.com/richardprobert/99efb44a5a04613475d1b2f67e837c20

In the example above, the bug manifests in ToString(), but is apparent in other operations too. The bug seems to be triggered by the combination of both an URN, and one that contains unicode characters.

I would have expected this to either work as expected, or throw some exception.

**Comment by davidsh** [comment]
You example is about .NET Framework.  Does this repro in .NET Core?  This repo is for reporting issues in .NET Core CoreFx code.

**Comment by davidsh** [comment]
cc: @rmkerr

**Comment by richardprobert** [comment]
My apologies, I was not aware of that. However, I can confirm that this is reproducible in a brand new .NET Core Console Application.

**Comment by davidsh** [comment]
Your repro code doesn't say what you expect the result of ToString() to be.  Can you clarify what "fails to parse URN" mean?  Also, the "URN" scheme is not something that System.Uri will treat special. It will be treated as a generic scheme  Can you point to an RFC section about the behavior you expect from the ToString() method?

**Comment by richardprobert** [comment]
Gist updated with expected and actual values.

**Comment by davidsh** [comment]
From Gist:

```c#
using System;

namespace UriTest
{
    class Program
    {
        static void Main(string[] args)
        {
            Uri uri = new Uri("urn:test/三维视图/3D.svf", UriKind.Absolute);
            string toString = uri.ToString();
        }
    }
}
```

Expected value of ToString(): "urn:test/三维视图/3D.svf"
Actual value of ToString(): "urn:test/三维视图/3D.svftest/三维视图/3D.svf"

**Comment by davidsh** [comment]
Since `UriKind.Absolute` was specified and the URI/IRI itself has no authority component, the string is really a UriKind.Relative.  So, this should probably have just thrown an exception.  Will investigate.  But like all URI bugs, we will consider app-compat before any changes to parsing behaviors.  Thx.

**Comment by richardprobert** [comment]
Please review updated Gist (or I can paste code in here if you prefer). I was trying to abbreviate the URI to clarify, but it appears I failed! Apologies. Updated Urn, plus expected and actual outcomes in comments.

No RFC, but as you can see behaviour is clearly inconsistant (and surely incorrect) from ASCII version.

**Comment by hotchkj** [comment]
>>>Since `UriKind.Absolute` was specified and the URI/IRI itself has no authority component, the string is really a `UriKind.Relative`.
I admit this confuses me. Are you saying that the NET `System.Uri` doesn't actually deal with URIs, but rather only with URIs that have authority components (I believe using RFC 3986 terminology this means _hiearchical_ URIs)?

Not my intent to be contrary, rather that on reading the [docs](https://msdn.microsoft.com/en-us/library/system.uri(v=vs.110).aspx) nothing actually makes this clear. Why shouldn't a user pass other schemes into `Uri`? URN is defined as a subset of URI - it is at least a little surprising that `System.Uri` would reject it outright. The remarks on that page even talk specifically about hierarchical URIs as though there are other kinds supported.

Said docs also state
>>>In the .NET Framework version 1.1, if the string specified to a constructor contains an unknown scheme and "c:\", the Uri class inserts "//" after the colon. For example, the URI xyz:c:\abc is converted to xyz://c:/abc. In the .NET Framework version 2.0, this behavior has been removed, and the example string is converted to xyz:c:/abc.

**Comment by davidsh** [comment]
>Since UriKind.Absolute was specified and the URI/IRI itself has no authority component, the string is really a UriKind.Relative.
I admit this confuses me. Are you saying that the NET System.Uri doesn't actually deal with URIs, but rather only with URIs that have authority components (I believe using RFC 3986 terminology this means hiearchical URIs)?

My comment was about the fact that the Uri constructor takes a UriKind parameter.  If the parameter is UriKind.AbsoluteOrRelative, then the string passed can be either an absolute or relative uri.  But if the parameter is UriKind.Absolute but the string passed in is a relative uri, then the constructor should throw exception.  And most uri's without authority portion in the string are relative URIs.

**Comment by rmkerr** [comment]
This is an issue we were having in .NET Framework too. I'll take a look at the code in .NET Core and see if I can port the change over.

**Comment by rmkerr** [comment]
Hey @richardprobert, can you give any more details on the version of .NET Core you were running when you encountered this issue? Looking into it, I have been unable to reproduce the issue with either the 2.0.4 release, or with the latest nightly build. This issue definitely exists in the current .NET Framework releases, and I'd really like to track down whether or not it occurs in .NET Core as well.

**Comment by richardprobert** [comment]
@rmkerr I think I was probably using .NET Core 1.0.x at the time. I can confirm that I can reproduce the issue with version 1.0.9 and 1.1.6, but **not** with version 2.0.5.

**Comment by rmkerr** [comment]
I'm glad to hear it seems to be resolved on 2.0.x! Thanks for taking the time to verify that. I'm going to close this bug, but if you run into anything similar in a new release please let me know.
